File: 17_Print1ToMaxOfNDigits/Print1ToMaxOfNDigits.py
import operator


class Solution(object):
    def Print1ToMaxOfNDigits(self, n):
        if n <= 0:
            return
        number = ['0'] * n
        for i in range(10):
            number[0] = str(i)
            self.__Print1ToMaxOfNDigitsRecursively(number, n, 0)

    def __PrintNumber(self, number):
        if (operator.eq(number, ['0'] * len(number))):
            return
        isBeginning0 = True
        nLength = len(number)

        for i in range(nLength):
            if isBeginning0 and number[i] != '0':
                isBeginning0 = False
            if not isBeginning0:
                print('%c' % number[i], end='')
        print('\t', end='')

    def __Print1ToMaxOfNDigitsRecursively(self, number, length, index):
        if index == length - 1:
            self.__PrintNumber(number)
            return
        for i in range(10):
            number[index + 1] = str(i)
            self.__Print1ToMaxOfNDigitsRecursively(number, length, index + 1)


solution = Solution()
solution.Print1ToMaxOfNDigits(4)

File: 17_Print1ToMaxOfNDigits/Print1ToMaxOfNDigits.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

bool Increment(char *number)
{
    bool isOverflow = false;
    int nTakeOver = 0;
    int nLength = strlen(number);
    for (int i = nLength - 1; i >= 0; i--)
    {
        int nSum = number[i] - '0' + nTakeOver;
        if (i == nLength - 1)
            nSum++;
        if (nSum >= 10)
        {
            if (i == 0)
                isOverflow = true;
            else
            {
                nSum -= 10;
                nTakeOver = 1;
                number[i] = '0' + nSum;
            }
        }
        else
        {
            number[i] = '0' + nSum;
            break;
        }
    }
    return isOverflow;
}

void PrintNumber(char *number)
{
    bool isBeginning0 = true;
    int nLength = strlen(number);

    for (int i = 0; i < nLength; i++)
    {
        if (isBeginning0 && number[i] != '0')
            isBeginning0 = false;
        if (!isBeginning0)
            printf("%c", number[i]);
    }
    printf("\t");
}

void Print1ToMaxOfNDigits(int n)
{
    if (n <= 0)
        return;
    char *number = new char[n + 1];
    fill(number, number + n, '0');
    number[n] = '\0';
    while (!Increment(number))
    {
        PrintNumber(number);
    }
    delete[] number;
}

int main()
{
    Print1ToMaxOfNDigits(4);
}